Fix part of #1719, part of #3709: Add build stats CI workflow

BenHenning opened the pull request:
## Explanation
Fixes part of #1719 and #3709

This PR introduces a new script & CI workflow for computing build stats to compare both AABs and universal APKs between develop and the changes in a given PR, as part of fixing #1719 (though this PR doesn't cover everything outlined in that PR). This information is then detailed and uploaded as a CI build artifact, and summarized & posted as a comment in the PR. Some details included in the summary report:
- APK file/download size differences
- Method count differences
- Feature/permission differences
- New/removed resources & assets

The script supports computing differences for multiple "profiles" at the same time, and the CI workflow has been set up to compute four:
1. dev
2. alpha
3. beta
4. GA

This workflow will be optional since it's very expensive to run (it has to assemble 8 builds, 6 of which are Proguarded). It also doesn't really need to be run in order to approve a PR, though reviewers may insist on waiting for large or suspicious changes 
(such as PRs introducing new dependencies) to ensure the actual affected changes are as expected.

In order to mitigate this expense, the CI workflow runs on a scheduled cron job off of develop across all open PRs and checks them in a group. It runs at most once per day (based on https://github.com/orgs/community/discussions/55768#discussioncomment-5941720) so multiple changes to a PR will be picked up with a single check in the next cron run. Currently, it will run even for a PR that hasn't changed since the last run (but this is something that can be improved in the future if it needs to be). It's being scheduled for 2:30am (02:30) UTC which seems to have a few specific benefits:
- Per GitHub documentation, initiating the workflow outside the start of the hour should reduce likelihood of cancellation (since the start of the hour tends to use the most resources): https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule.
- This corresponds to 7:30pm PT, 2:30am GMT, 8:00am IST, 5:30am EAT, and 12:30pm AEST (just as a basis for a different part of the world). It's actually a very nice time that shouldn't overlap with almost any development in main locations around the world, so it hopefully won't impact Oppia organization GitHub CI resources.

The example output from these workflows can be observed in a few places:
- Later in this PR (for back when the PR was configured to run the new workflow per PR change).
- In #4261 which demonstrates the large math PRs and how they changed the builds back before those were merged.
- https://github.com/BenHenning/oppia-android/actions/workflows/stats.yml and https://github.com/BenHenning/oppia-android/pulls (specifically: https://github.com/BenHenning/oppia-android/pull/14, https://github.com/BenHenning/oppia-android/pull/13, https://github.com/BenHenning/oppia-android/pull/12) which demonstrates the workflow running correctly from a scheduled cron (https://github.com/BenHenning/oppia-android/actions/runs/9232187176) and posting the updates to open PRs.

Beyond that, implementing this utility involved several significant changes to various systems, including the build graph:
- Three new utilities were added for the script: Aapt2Client, ApkAnalyzerClient, and BundleToolClient. Each of these correspond to Android CLI utilities, but each required special considerations:
  - Aapt2Client requires direct access to the Android SDK, but fortunately android_sdk_repository exposes this as a target so it's trivial to pass it in & call it. Some build information is needed, too (see next outer point).
  - ApkAnalyzerClient couldn't use the apkanalyzer CLI contained within the SDK since it's not exported by android_sdk_repository. Instead, we needed to depend on the CLI's internal implementation library (which I suspect is what Android Studio probably uses for its own APK Analyzer tool). This required some new implementation.
  - BundleToolClient fortunately can call right into the bundle tool library that we use when building AABs, but unfortunately that tool appears to not be designed to be called multiple times in the same process. Because Java doesn't support forking, we actually needed to fake a fork function by starting a new Java process using the current process's classpath in order to re-run bundle tool for each needed routine. Additionally, bundle tool required https://github.com/oppia/archive-patcher (which needed new BUILD files since it only supported Gradle building previously) and a non-Android version of Guava (see below for the changes this has caused).
- A new build_vars.bzl was introduced to define the build SDK & build tools versions (this is done in a way where they can actually be passed to the new script's utilities since it needs to access aapt2).
- rules_kotlin had a bug where resources wouldn't be pulled in properly for kt_jvm_library (see https://github.com/bazelbuild/rules_kotlin/issues/281), but this was mitigated in a previous PR by upgrading rules_kotlin past alpha 2.
- The new functionality required the JRE-compatible version of Guava (over the Android-constrained library used in the codebase today), but this introduces a one-version issue. The solution ended up being isolating the JRE-compatible Guava library to its own library with a slightly hacky direct reference to it in BundleToolClient. Some of the other attempts at solving this resulted in some Maven reference cleanups in existing script documentation. This functionality will be improved in downstream PRs, but other attempts that were originally made to isolate this cleanly were:
  - Introduce multiple maven_install files and isolate dependencies into: production, tests, scripts. This has a number of nice benefits (more correct licenses and faster Maven dependency fetches for production), but it results in very tricky one-version violations for test targets that cross dependencies between production and tests.
  - Isolated maven_install just for scripts. This is closer to the solution we'll want long-term, but it was too much complexity to fully introduce in this PR so it's been reworked into a downstream PR that can focus on cleaning up third-party dependency management across the whole codebase.

This PR is introducing a few new dependencies that, in turn, pull in a *bunch* of transitive dependencies. These are all due to the new ``apkanalyzer`` dependency. While it will affect licenses for this specific PR, once third-party dependencies for scripts are cleaned up in a downstream PR they will be moved out (since they are script-only dependencies).

Separately, also note that the AAPT2 utility requires stdout to be processed continuously in order for the process to finish. This was one of the primary reasons CommandExecutorImpl was reworked in #4929.

For testing: most of the changes in this PR have been extensively manually tested. However, the new utilities are lacking significant automated tests. Since this utility is a nice-to-have for the rest of the Bazel PR chain, it's being prioritized to be merged in spite of lacking code coverage. #4971 has been filed to track adding these missing tests in the long-term.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- This only affects CI workflows & the build system. Technically, some dependency changes in the build system could have UI effects, but there should be no such changes in this PR.

BenHenning committed:
Add script for computing AAB/APK differences.

This script primarily focuses on high-level differences, including:
- Sizes (file + download)
- New/old files, configurations, resource changes
- Method counts

BenHenning committed:
Add workflow to show build stats.

This workflow is set up to build multiple binaries for both develop &
the local branch, then analyze and generate reports that are uploaded as
artifacts and added as a comment to the PR body.

This doesn't make a guarantee about exact differences since GitHub uses
a detached HEAD for the commit that may not exactly match.

This also introduces a new workflow for stats (that will initially only
contain optional workflows since they may be very expensive/time
consuming).

BenHenning committed:
Add missing new build vars.

These allow some SDK properties to be shared across different
dependencies.

BenHenning committed:
Tests + docs + build changes.

This updates the script to no longer require an Android SDK path to be
passed in since it now directly pulls in AAPT2 and directly performs the
work of apkanalyzer using its internal library rather than calling
through to the CLI library (since it's not exposed via
android_sdk_repository).

This also updates the report output to compute full comparisons for all
APKs (rather than just the universal APK) and include it as part of the
full summary (so that only 2 files are ever written regardless of the
number of profiles being compared).

This adds placeholder tests & documentation for all new utilities.

This commit includes substantial changes to the build system:
- A new dependency on https://github.com/oppia/archive-patcher
- A split between production, test, and script Maven dependencies (where
the license tools have been updated to only operate on production
dependencies since those are the ones being shipped to users)
- Two versions of Guava are now being pulled in since the scripts
require a JRE version of Guava rather than the Android version used in
production & test classes
- rules_kotlin has been upgraded to beta 3 (from alpha 2), but the
Kotlin version itself is stuck on 1.4 (to ensure interoperability with
previous builds). While this change doesn't upgrade Kotlin, it does
allow us to leverage better tooling per changes in rules_kotlin,
including enforcing strict deps.

BenHenning committed:
Fix CommandExecutorImpl's timeout feature.

Also, since I forgot to add these notes in the previous commit: the
executor has also been changed in this PR to support applications which
block on stdout/stderr being consumed.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into add-build-stats-ci-workflow

Conflicts:
	third_party/maven_prod_install.json

BenHenning committed:
Fix post-merge conflicts.

BenHenning committed:
Fix workflow name.

BenHenning committed:
Fix workflow names & IDs.

Remove duplicated workflow.

BenHenning committed:
Temporarily disable workflows -> faster iteration.

BenHenning committed:
Add Proguard exemption.

I'm not entirely sure why this is needed yet.

BenHenning committed:
Remove broken & unneeded copy line.

BenHenning committed:
Fix variable reference for comment body.

BenHenning committed:
Add more collapsing in comment.

The current comment is a bit too long, so adding additional collapsing
per-flavor.

BenHenning committed:
Only isolate special script deps.

This reverts the split prod/test deps since it results in one version
issues.

This commit also fixes the nice byte formatting in the report not
working correctly for negative sizes.

BenHenning committed:
Re-enable CI workflows.

BenHenning commented:
Message so that this PR is not considered stale. I'm not actively working on it yet, but I don't want it to be closed.

BenHenning commented:
Still planning to revive this (hopefully sometime next week).

BenHenning commented:
This is still active.

BenHenning committed:
Merge branch 'develop' into add-build-stats-ci-workflow

Conflicts:
	WORKSPACE
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes & wording improvements.

BenHenning committed:
Upgrade to Kotlin 1.6.

This PR resolves *all* Maven conflicts and Proguard issues by ensuring
everything is using the latest Kotlin runtime & libraries.

Tests have not been verified as building yet (they probably won't due to
visibility changes & new behaviors/API changes introduced in Kotlin 1.5
and 1.6).

This is needed as part of checking #2173 for build stats (which will
happen on another, temporary branch).

BenHenning committed:
Fix guava_jre deps & in bundle tool client.

BenHenning committed:
Buildifier lint fixes.

BenHenning commented:
(Note that the above stats include removing kotlin-reflect which I think added a lot of methods)

BenHenning committed:
Migrate to rules_jvm_external 5.1.

BenHenning committed:
Fix JVm dep build warnings.

BenHenning committed:
Fix Maven license check & update the list.

The check & related scripts required fairly involved reworking since the
Maven install file (from which it sources Maven URL context) changed in
format as part of the upgrade for rules_jvm_external. This has actually
led to what seems to be more correct analysis of libraries that the
build depends on, so more licenses have been added to the
maven_dependencies.textproto tracking file.

One unused Crashlytics dependency was removed since it was referencing
an old license that doesn't exist anymore (and the artifact should be
replaced in full by more recent Firebase Crashlytics dependencies that
we are already using).

BenHenning committed:
Fix issues with compute_affected_tests.

This addresses an underlying bug with the command executor that can, in
some cases, break compute_affected_tests. It also refines some of its
internal mechanisms for much better performance on expensive PRs.

It also prepares the base support needed for merge queues, but the
CI workflows aren't being updated in this change.

BenHenning committed:
Use base SHA for computing affected tests.

This prepares for merge queues (but doesn't quite configure the workflow
for them--that will happen in a different PR), and improves how tests
are computed for stacked PRs.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	scripts/src/java/org/oppia/android/scripts/common/BUILD.bazel
	scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt
	scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesRetriever.kt
	scripts/src/java/org/oppia/android/scripts/maven/GenerateMavenDependenciesList.kt
	scripts/src/javatests/org/oppia/android/scripts/license/MavenDependenciesListCheckTest.kt
	scripts/src/javatests/org/oppia/android/scripts/license/MavenDependenciesRetrieverTest.kt
	scripts/src/javatests/org/oppia/android/scripts/maven/GenerateMavenDependenciesListTest.kt

BenHenning committed:
Post-merge lint fix.

BenHenning committed:
Fix updated TODOs.

Also, update TODO check script to have nicer output, and support
generating the exemption textproto file for easier updates in the
future.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Fix Dagger structure.

This moves the codebase to using the recommended single top-level Dagger
library rather than replicating it in a bunch of different places.

BenHenning committed:
Upgrade Dagger version.

This is needed for downstream work. It also includes ensuring that Guava
JRE can never be used (since only Android should ever be referenced by
the production app build).

BenHenning committed:
Fix test builds that depend on Dagger generation.

BenHenning committed:
Update Maven dependency licenses.

BenHenning committed:
Fix broken test & binary builds.

BenHenning committed:
Add support for Kotlin 1.6.

There's some cleanup work needed beyond this, but this is the core
change to introduce Kotlin 1.6 support (at least for dev builds).

Moshi needed to be upgraded due to a metadata incompatibility when moving
over to Kotlin 1.6.

BenHenning committed:
Correctly interact with rules_kotlin.

rules_kotlin moved its imports into more structured bzl files to load,
so this updates all references in the codebase to point to the correct
locations (which removes debug warnings that show up on the CLI).

BenHenning committed:
Fix upgrade by downgrading Moshi to 1.13.

Moshi 1.14 pulls in kotlin-reflect 1.7.0 which isn't compatible with the
rules_kotlin version we need for Bazel 4.x. Relatedly, this downgrades
rules_kotlin to 1.5.0, but it fortunately keeps all other changes needed
for 1.7.1 (which will be used in a later PR).

Some code fixes were needed, too, for unknown reasons (since the build
should've been using Kotlin 1.6 before). Either way, these fixes seem
reasonable.

BenHenning committed:
Turn on warnings as errors for Kotlin.

Fixed all warnings that the compiler warned about.

Removed ViewModelProvider & fixed state leaking entirely by moving away
from Jetpack's ViewModel as a base class (since we aren't correctly
using that correctly).

BenHenning committed:
Fix Java build & other warnings.

Enables Java warnings-as-errors, though this doesn't yet apply to
kapt-generated code (such as the code from Dagger), but those warnings
were still manually fixed.

This also fixes a small import warning in a proto file, and warnings
when building oppia_dev_kitkat (by updating the main dex list, but it's
likely the build doesn't work anymore, anyway, and it's hard to test
locally).

BenHenning committed:
Try to fix AndroidX databinding compile race.

There's a race condition when building large numbers of app tests
simultaneously that can lead to build failures. While the CI runs are
resilient now to this failure, it'd be better to try and fix it. This
removes the last non-AndroidX dependency from the codebase with hopes
that it helps reduce the likelihood of the error (though there are no
dependencies on it, so it's unlikely).

BenHenning committed:
Fix static check failures.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

Conflicts:
	WORKSPACE
	app/build.gradle
	app/src/main/java/org/oppia/android/app/viewmodel/ViewModelBridgeFactory.kt
	config/proguard/kotlin-proguard-rules.pro
	domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/AnalyticsController.kt
	scripts/assets/maven_dependencies.textproto
	scripts/src/java/org/oppia/android/scripts/common/BUILD.bazel
	scripts/src/java/org/oppia/android/scripts/common/CommandExecutorImpl.kt
	scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesRetriever.kt
	scripts/src/java/org/oppia/android/scripts/maven/model/MavenListDependencies.kt
	scripts/src/java/org/oppia/android/scripts/maven/model/MavenListDependency.kt
	scripts/src/java/org/oppia/android/scripts/maven/model/MavenListDependencyTree.kt
	third_party/BUILD.bazel
	third_party/maven_install.json
	third_party/versions.bzl
	tools/kotlin/BUILD.bazel
	utility/build.gradle
	utility/src/main/java/org/oppia/android/util/parser/image/UrlImageParser.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Fix Proguard builds (temp) and static failures.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

BenHenning committed:
Fix CI breakages, including Gradle.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

Conflicts:
	build.gradle
	data/build.gradle
	domain/build.gradle
	third_party/maven_install.json

BenHenning committed:
Post-merge dep fixes + test fix.

BenHenning committed:
Undo unnecessary changes.

BenHenning committed:
Fix broken CI checks.

Specifically:
- The TODO check was fixed by reformatting a TODO comment.
- The Proguard build was fixed by upgrading kotlinx.coroutines to use a
  version compatible with Kotlin 1.6, as well as adding a Proguard
  dontwarn directive for one class that can't execute on Android.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

Conflicts:
	WORKSPACE
	config/proguard/kotlin-proguard-rules.pro
	third_party/BUILD.bazel
	third_party/maven_install.json
	third_party/versions.bzl

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Address self-review comments.

BenHenning committed:
Fix broken tests.

For some reason, the old kotlinx-coroutines-test was causing a
codebase-wide incompatibility with the 1.6.x kotlinx-coroutines-core
library, so this updates that version.

It also updates Gradle dependencies for consistency, and to fix the
Gradle builds.

And, this fixes the IntelliJ Bazel plugin syncing issue when using
kt_jvm_import (for kotlinx-coroutines-core-jvm).

BenHenning committed:
Address TODOs as part of the migration.

This addresses TODOs for #4119 and #4120. It also updates TODOs on #3715
to #4463, instead, since these apparently weren't correctly originally.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

Conflicts:
	third_party/BUILD.bazel
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Update build stats workflow.

This changes the following:
- Removes the alpha_kitkat build (since it's no longer important).
- Adds build checks for beta & GA builds.
- Updates the logic to run against the base commit rather than develop
  for better interoperability with chained PRs.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

BenHenning committed:
Fix breakage from develop.

The new proto target isn't used anywhere so this was missed.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Conflicts:
	app/src/main/res/layout-land/profile_chooser_profile_view.xml
	app/src/main/res/layout-sw600dp-port/profile_chooser_profile_view.xml
	app/src/main/res/layout/profile_chooser_profile_view.xml

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning committed:
Fix hanging CoroutineExecutorService test.

I'm not sure of the underlying problem, but we've run into issues with
the custom Deferred-to-Future conversion mechanism. This replaces that
with an existing kotlinx Deferred-to-Guava ListenableFuture mechanism
that is used elsewhere in the codebase and it seems to fix the stability
issue.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning committed:
Permant-ify temporary TODOs for new tests.

Also, fix some oddities in the API contract for functions returning
sizes as strings rather than longs.

Issue #4971 is tracking the long-term effort to add tests for the new
utilities being introduced by this branch's PR.

BenHenning committed:
Upgrade runner for build stats.

This ensures that the workflow can actually run since GitHub doesn't
support the old 18.04 Ubuntu images anymore (or they are at least
significantly limited in availability).

BenHenning committed:
Fix broken Gradle build.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning commented:
Still active.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	scripts/assets/test_file_exemptions.textproto
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Conflicts:
	app/src/main/java/org/oppia/android/app/translation/BUILD.bazel
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Conflicts:
	.github/CODEOWNERS
	app/src/main/java/org/oppia/android/app/options/AppLanguageActivity.kt
	app/src/main/java/org/oppia/android/app/options/AppLanguageActivityPresenter.kt
	app/src/main/java/org/oppia/android/app/options/AppLanguageFragment.kt
	app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/options/OptionsActivity.kt
	app/src/main/java/org/oppia/android/app/options/OptionsActivityPresenter.kt
	app/src/main/java/org/oppia/android/app/options/OptionsFragmentPresenter.kt
	app/src/sharedTest/java/org/oppia/android/app/options/AppLanguageActivityTest.kt
	app/src/sharedTest/java/org/oppia/android/app/options/AppLanguageFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/options/OptionsFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/test/java/org/oppia/android/app/translation/AppLanguageResourceHandlerTest.kt
	app/src/test/java/org/oppia/android/app/utility/math/MathExpressionAccessibilityUtilTest.kt
	scripts/src/java/org/oppia/android/scripts/build/FilterPerLanguageResources.kt
	third_party/maven_install.json
	utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Fix a variety of dev platform-specific issues.

These issues were found after I started using a new development
environment.

BenHenning committed:
Tidy some things up, and revert fragment test.

ProfileAndDeviceIdFragmentTest had been updated to use a newer fragment
initialization pattern, but that's no longer needed and seems to be
causing what appears to be timing discrepancies between local dev and
CI.

BenHenning committed:
Fix test behavior inconsistency in test.

The issue ultimately arose from test parameters being initialized after
they're needed in the launched UI. This type of change was tried earlier
in the branch, but reverted since it didn't seem necessary. It is,
however, necessary when there are environment differences (e.g. local
vs. CI) or when running certain tests individually.

Due to the difficulty in finding this issue, ActivityScenarioRule has
been added as a prohibited pattern in the static regex checks (along
with ActivityTestRule since that's deprecated and discouraged, anyway).

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Fix broken ProfileAndDeviceIdFragmentTest test.

The test was suffering from some proto encoding inconsistencies that
seem to occur between some development machines vs. on CI. The fix
improves the test's robustness by extracting the raw encoded string,
verifying that the other outputs in the intent message correctly
correspond to that string, and that the string (as a parsed proto)
contains the correct values. As a result, the test no longer depends on
a hardcoded encoding value to be present for verification. This does
result in a bit more logic than is generally good to have in a test (and
it lengthened the test code quite a bit), but it seems necessary in this
particular case.

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Post-merge fix.

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/ci/ComputeAffectedTestsTest.kt
	scripts/src/javatests/org/oppia/android/scripts/common/GitClientTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/ci/ComputeAffectedTestsTest.kt
	scripts/src/javatests/org/oppia/android/scripts/common/GitClientTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Some follow-up post-merge fixes.

This doesn't address all timeouts being observed in scripts. More
analysis may be needed.

BenHenning committed:
Make TodoOpenCheck locally runnable.

This does a bunch of other small things, too, but the main difference
here is introducing support for interacting with GitHub via its REST API
rather than requiring the user to use the 'gh' tool locally (though the
gh CLI tool is still needed for maintaining authentication access).

BenHenning committed:
Simplify GitHubIssue & some static fixes/docs.

Tests still need to be added, and some cleanup is needed once the
branch's base is adjusted.

BenHenning committed:
Add ScriptBackgroundCoroutineDispatcher.

This is a script-specific dispatcher which will allow for better
asynchronous support in upcoming PRs (especially for command execution).
This change serves to prepare for those changes.

BenHenning committed:
Some more robustness fixes.

The main change here is ensuring that Bazel 4.0.0 is used & bzlmod
disabled in newer versions of Bazel when running operations in a test
Bazel environment.

This commit also introduces some more timing tweaks on CommandExecutor
for some tests, though these only affect very specific tests (as many
script tests directly call a script's main() function and thus don't
overwrite its executor behavior).

This commit attempted to introduce "--batch" mode to runs, but the
isolation didn't actually seem to improve stability and, instead,
substantially slowed down some of the tests.

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

Conflicts:
	scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt
	scripts/src/javatests/org/oppia/android/scripts/ci/ComputeAffectedTestsTest.kt
	scripts/src/javatests/org/oppia/android/scripts/common/BazelClientTest.kt

BenHenning committed:
Post-merge fixes.

Plus, actually makes use of the new script background dispatcher in
CommandExecutorImpl to make the new wiring make more sense (though its
real utility will come in a follow-up PR).

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Post-merge fixes & adjustments.

This makes the TODO open check script runnable using the script
background dispatcher, and fixes some tests so that they pass now.

More documentation and testing work is still needed to finalize this PR.

BenHenning committed:
Fix BUILD file lint issue.

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Static fix checks.

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt
	scripts/src/java/org/oppia/android/scripts/common/BazelClient.kt
	scripts/src/java/org/oppia/android/scripts/common/CommandExecutorImpl.kt
	scripts/src/java/org/oppia/android/scripts/common/GitClient.kt
	scripts/src/java/org/oppia/android/scripts/todo/BUILD.bazel
	scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt
	scripts/src/javatests/org/oppia/android/scripts/ci/ComputeAffectedTestsTest.kt
	scripts/src/javatests/org/oppia/android/scripts/common/ScriptBackgroundCoroutineDispatcherTest.kt
	scripts/src/javatests/org/oppia/android/scripts/testing/TestGitRepositoryTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' of github.com:oppia/oppia-android into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	scripts/assets/test_file_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Add smoke tests for instr. binaries & tests.

BenHenning committed:
Some minor refactoring for readability.

BenHenning committed:
Fixed broken instrumentation builds.

This was done by removing the //testing dependency and, instead, having
instrumentation targets depend on the direct module within //testing
that they need to build. This module & its corresponding implementation
binding (and tests) needed to be moved out of //testing and into their
own /firebase package.

BenHenning committed:
Merge branch 'develop' into fix-instrumentation-build-failure

BenHenning committed:
Add missing CODEOWNERS line.

BenHenning committed:
Remove old files from CODEOWNERS.

BenHenning committed:
Merge branch 'fix-instrumentation-build-failure' into fix-platform-specific-issues

BenHenning committed:
Add missing tests for TestBlazeWorkspace changes.

Also includes some minor clean-ups within TestBazelWorkspaceTest.

BenHenning committed:
Add missing Firebase auth tests.

BenHenning committed:
Merge branch 'fix-instrumentation-build-failure' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Post-merge fixes for previously missed tests.

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'develop' into fix-instrumentation-build-failure

BenHenning committed:
Merge branch 'fix-instrumentation-build-failure' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'develop' into fix-instrumentation-build-failure

BenHenning committed:
Merge branch 'fix-instrumentation-build-failure' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

BenHenning committed:
Improve new test robustness.

Required upgrade of kotlinx-coroutines-core.

BenHenning committed:
Fix updated Maven license deps.

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Update scripts/src/javatests/org/oppia/android/scripts/common/GitClientTest.kt

Co-authored-by: Adhiambo Peres <59600948+adhiamboperes@users.noreply.github.com>

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

Conflicts:
	third_party/maven_install.json

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Add tests for GitHubClient & TodoOpenCheck.

Add new FakeCommandExecutor utility + tests.

Add missing GitHub workflow update.

BenHenning committed:
Merge branch 'develop' into fix-platform-specific-issues

BenHenning committed:
Merge branch 'fix-platform-specific-issues' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Add missing docs.

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	scripts/assets/maven_dependencies.textproto
	third_party/maven_install.json

BenHenning committed:
Update Maven dep list checker & updater scripts.

Updates the utilities for the Maven dep list & updater scripts to have
much more robustness when trying to verify the correct repository URLs
for artifact coordinates, and when downloading artifact POM files. The
main robustness changes come in the form of retries and smarter
parallelization and error checking.

These changes were necessary due to one of my workstations frequently
flaking out on large numbers of URL blocks during verification, and this
helps considerably.

BenHenning committed:
Merge branch 'develop' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Fix broken TodoOpenCheck tests & check exemptions.

For the former, this mainly involved isolating the gh auth tool
dependency in the same way as GitHubClientTest. This improves
environment independence and test performance.

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

Conflicts:
	scripts/assets/todo_open_exemptions.textproto
	scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Fix broken checks & failing static checks.

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Includes some manual fixes due to ":dagger" being moved to top-level.

Conflicts:
	app/BUILD.bazel
	app/src/main/java/org/oppia/android/app/notice/testing/BUILD.bazel
	app/src/sharedTest/java/org/oppia/android/app/notice/BUILD.bazel
	domain/src/main/java/org/oppia/android/domain/onboarding/BUILD.bazel
	domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/BUILD.bazel
	domain/src/test/java/org/oppia/android/domain/exploration/BUILD.bazel
	domain/src/test/java/org/oppia/android/domain/onboarding/BUILD.bazel
	domain/src/test/java/org/oppia/android/domain/oppialogger/analytics/BUILD.bazel
	scripts/assets/maven_dependencies.textproto
	third_party/maven_install.json
	third_party/versions.bzl
	utility/src/main/java/org/oppia/android/util/logging/firebase/BUILD.bazel
	utility/src/test/java/org/oppia/android/util/logging/firebase/BUILD.bazel

BenHenning committed:
Post merge build, test, and static check fixes.

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Includes some additional clean-up changes in ClickableAreasImage.

Conflicts:
	app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt
	app/src/main/java/org/oppia/android/app/testing/InputInteractionViewTestActivity.kt
	app/src/main/java/org/oppia/android/app/utility/ClickableAreasImage.kt
	app/src/main/res/layout/activity_input_interaction_view_test.xml
	app/src/main/res/layout/nav_header_navigation_drawer.xml
	app/src/main/res/layout/profile_picture_activity.xml
	domain/build.gradle
	instrumentation/oppia_instrumentation_test.bzl
	scripts/assets/maven_dependencies.textproto
	testing/build.gradle
	third_party/maven_install.json
	third_party/versions.bzl

BenHenning committed:
Merge branch 'develop' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Conflicts:
	utility/src/main/java/org/oppia/android/util/locale/BUILD.bazel

BenHenning committed:
Some post-merge fixes.

More are needed yet to address all of the Kotlin warnings (which are now
treated as errors).

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

BenHenning committed:
Silence deprecation warnings.

These will need to be fixed properly before the PR is ready for review.

BenHenning committed:
Merge branch 'develop' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Address reviewer comment.

seanlip committed:
Merge branch 'develop' into introduce-better-script-execution-support

BenHenning committed:
Merge commit 'e2f94e41e4cba3bc5b5f3af65f7c43a6543b4c52' into introduce-better-script-execution-support

BenHenning committed:
Merge branch 'introduce-better-script-execution-support' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'develop' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'develop' into make-todo-check-locally-runnable

BenHenning committed:
Merge branch 'make-todo-check-locally-runnable' into upgrade-compute-affected-tests

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

Conflicts:
	WORKSPACE
	third_party/maven_install.json

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Conflicts:
	domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/BUILD.bazel
	third_party/maven_install.json
	utility/src/test/java/org/oppia/android/util/logging/BUILD.bazel

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Conflicts:
	app/src/main/java/org/oppia/android/app/testing/InputInteractionViewTestActivity.kt
	app/src/main/res/layout/activity_input_interaction_view_test.xml
	app/src/sharedTest/java/org/oppia/android/app/customview/interaction/MathExpressionInteractionsViewTest.kt
	domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/AnalyticsController.kt
	third_party/maven_install.json

BenHenning committed:
Address self-review comments.

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Address review comment.

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

Conflicts:
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-kotlin1.6

Conflicts:
	domain/src/main/java/org/oppia/android/domain/auth/BUILD.bazel
	domain/src/main/java/org/oppia/android/domain/oppialogger/survey/BUILD.bazel
	domain/src/main/java/org/oppia/android/domain/survey/BUILD.bazel
	testing/src/main/java/org/oppia/android/testing/firebase/BUILD.bazel
	third_party/maven_install.json

BenHenning committed:
Fix NetworkLoggingInterceptorTest.

BenHenning committed:
Fix more tests & upgrade Robolectric.

Fixing one of the test warnings required upgrading to Robolectric 4.9
(to add support for SDK 33 which the app has targeted for the past ~6
months).

BenHenning committed:
Fix more broken tests.

BenHenning committed:
Undo Robolectric upgrade.

Robolectric upgrade is hard: it requires upgrading Espresso in order to
work around https://github.com/robolectric/robolectric/issues/7367 and
https://github.com/robolectric/robolectric/issues/6593. However, newer
Espresso makes use of new manifest features that require upgrading
Bazel: https://github.com/bazelbuild/bazel/issues/14285. Because
upgrading Kotlin is a precursor to the Bazel upgrade, a workaround for
the single test that required the Robolectric upgrade (due to using a
deprecated API from SDK 33 without being able to actually use SDK 33 in
the test) is the best approach here until Robolectric can actually be
upgraded in the future.

BenHenning committed:
Upgrade to rules_kotlin 1.5.0 beta 3.

This is in preparation for the codebase-wide upgrade to Kotlin 1.6.

BenHenning committed:
Merge branch 'upgrade-to-newer-rules-kotlin' into upgrade-to-kotlin1.6

Conflicts:
	WORKSPACE
	scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt
	third_party/versions.bzl

BenHenning committed:
Fix build-time Kotlin & Java warnings.

This pulls in non-configuration changes from #4937 that are compatible with the
current Kotlin 1.4.x. Kotlin 1.5 and 1.6 changes will need to come as part of a
downstream PR.

BenHenning committed:
Fix a few follow-up issues.

These are mainly reverting a few more Kotlin 1.5/1.6-specific API
changes.

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/customview/interaction/MathExpressionInteractionsViewTest.kt
	app/src/test/java/org/oppia/android/app/translation/AppLanguageResourceHandlerTest.kt
	data/src/test/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptorTest.kt
	testing/src/main/java/org/oppia/android/testing/junit/ParameterizedMethod.kt
	utility/src/test/java/org/oppia/android/util/logging/ConsoleLoggerTest.kt

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

Conflicts:
	WORKSPACE
	scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesRetriever.kt
	third_party/maven_install.json

BenHenning committed:
Post-merge fixes.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Merge branch 'develop' into upgrade-compute-affected-tests

BenHenning committed:
Merge branch 'upgrade-compute-affected-tests' into update-rules-jvm-external

BenHenning committed:
Follow-up variable rename for clarity.

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-newer-rules-kotlin

BenHenning committed:
Merge branch 'upgrade-to-newer-rules-kotlin' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Address static check failures.

This mainly involved updating CODEOWNERS & fixing a bunch of TODO
references for now-marked deprecations.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning committed:
Move forward TODO fixes from #4937.

These changes should have been in this branch originally.

BenHenning committed:
Remove unused view model bridge classes.

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

Conflicts:
	app/src/main/java/org/oppia/android/app/viewmodel/BUILD.bazel

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning committed:
Rework CI workflow.

This sets up the CI for build stats to run on a cron job across all PRs
rather than for each PR change.

BenHenning commented on WORKSPACE:
```diff
@@ -221,3 +236,7 @@ http_jar(
     ],
 )
+
+load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")
+
+bazel_skylib_workspace()
```
Verify whether this is actually needed.

BenHenning replied:
Done--it doesn't seem to be needed.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/Aapt2Client.kt:
```diff
@@ -0,0 +1,65 @@
+  private fun executeApkAnalyzerCommand(vararg arguments: String): List<String> {
+    val result = commandExecutor.executeCommand(workingDirectory, aapt2Path, *arguments)
+    check(result.exitCode == 0) {
+      "Expected non-zero exit code (not ${result.exitCode}) for command: ${result.command}." +
```
Should be zero, not non-zero.

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/BazelClient.kt:
```diff
@@ -84,8 +84,8 @@ class BazelClient(private val rootDirectory: File, private val commandExecutor:
   /**
-   * Returns the list of direct and indirect maven third-party dependencies on which the specified
-   * binary depends.
+   * Returns the list of direct and indirect production Maven third-party dependencies on which the
```
Remove 'production'.

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/BUILD.bazel:
```diff
@@ -88,3 +99,13 @@ kt_jvm_library(
 )
+
+genrule(
+    name = "sdk_info_properties",
```
Nit: should start with '_'.

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/AndroidBuildSdkProperties.kt:
```diff
@@ -0,0 +1,34 @@
+  val buildSdkVersion: Int
+    get() = androidSdkProperties.getExpectedProperty("build_sdk_version").toInt()
+
+  /** The version of Android built tools used for building. */
```
s/built/build

BenHenning replied:
Used CLI instead to avoid redundancy.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/common/AndroidBuildSdkProperties.kt:
```diff
@@ -0,0 +1,34 @@
+    // Note that this path must start with '/' since Bazel places the properties file not properly
+    // relative to this class (since it's nested under a separate src directory; this may be fixable
+    // in the future if script sources were ever moved to a top-level src directory).
+    private const val SDK_INFO_PROPERTIES_PATH =
+      "/scripts/src/java/org/oppia/android/scripts/common/sdk_info.properties"
```
Fix this; we can use class-relative resources, instead, as done with protos for other scripts.

BenHenning replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/maven/GenerateMavenDependenciesListTest.kt:
```diff
@@ -1342,7 +1342,7 @@ class GenerateMavenDependenciesListTest {
   }
 
-  /** Helper function to write a fake maven_install.json file. */
+  /** Helper function to write a fake maven install manifest file. */
```
s/maven/Maven for consistency.

BenHenning replied:
Done.

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/common/AndroidBuildSdkPropertiesTest.kt:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  @Test
+  fun testBuildToolsVersion_isTheVersion() {
```
```suggestion
  fun testBuildToolsVersion_isTheCorrectVersion() {
```

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/ApkAnalyzerClient.kt:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  // Based on the apkanalyzer CLI implementation.
+  private fun walkParseTree(node: DefaultMutableTreeNode?): List<String> {
```
Fix method ordering in this class.

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/ApkAnalyzerClient.kt:
```diff
@@ -0,0 +1,105 @@
+
+  private fun collectZipEntries(
+    inputZipFile: String,
+    predicate: (String) -> Boolean
```
s/predicate/namePredicate

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/ApkAnalyzerClient.kt:
```diff
@@ -0,0 +1,105 @@
+        .asSequence()
+        .filter { predicate(it.name) }
+        .associateBy { it.name }
+        .mapValues { (_, entry) -> zipFile.getInputStream(entry).readBytes() }
```
Close stream.

BenHenning replied:
Doesn't actually need to be closed since it's part of a ZipFile class that's already being closed (per ZipFile docs).

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/BundleToolClient.kt:
```diff
@@ -0,0 +1,126 @@
+ * General utility for interfacing with bundletool in the local system at the specified working
+ * directory path.
+ *
+ * @property commandExecutor the [CommandExecutor] to use when accessing bundletool
```
Add missing props.

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/Aapt2Client.kt:
````diff
@@ -0,0 +1,65 @@
+ * data = ["@androidsdk//:aapt2_binary"]
+ * ```
+ *
+ * @property buildToolsVersion the version of Android build tools installed & that should be used.
````
Add missing props.

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/BundleToolClient.kt:
```diff
@@ -0,0 +1,126 @@
+        *arguments
+      )
+    check(result.exitCode == 0) {
+      "Expected non-zero exit code (not ${result.exitCode}) for command: ${result.command}." +
```
Expected zero, not non-zero.

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/ComputeAabDifferences.kt:
```diff
@@ -0,0 +1,797 @@
+    val universalApkWithoutChanges = computeUniversalApk(aabWithoutChangesPath, destWithoutChanges)
+    val universalApkWithChanges = computeUniversalApk(aabWithChangesPath, destWithChanges)
+
+    val (masterApkWithoutChanges, splitApksWithoutChanges) =
```
Use better language here.

BenHenning replied:
Done.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/apkstats/ComputeAabDifferences.kt:
```diff
@@ -0,0 +1,797 @@
+     * @param itemLimit the max number of items to include in expanded lists
+     * @param itemize whether to expand lists of items
+     */
+    fun writeTo(stream: PrintStream, itemize: Boolean, itemLimit: Int): Boolean {
```
Document return value.

BenHenning replied:
Done.

BenHenning reviewed (commented):
Self-reviewed all changes.

BenHenning committed:
Address all reviewer comments.

BenHenning reviewed (commented):
Self-reviewed latest changes.

BenHenning commented:
@adhiamboperes PTAL.

BenHenning committed:
Pull in other warning fixes from #4886.

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Pulled forward Gradle change from #4886.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning committed:
Merge branch 'develop' into update-rules-jvm-external

BenHenning committed:
Merge branch 'develop' into update-rules-jvm-external

BenHenning committed:
Merge branch 'update-rules-jvm-external' into update-dagger-structure

BenHenning committed:
Merge branch 'develop' into update-dagger-structure

BenHenning committed:
Merge branch 'update-dagger-structure' into upgrade-to-newer-rules-kotlin

BenHenning committed:
Merge branch 'upgrade-to-newer-rules-kotlin' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Down version to 1.6.10.

This is to ensure Jetpack Compose compatibility.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

Conflicts:
	third_party/maven_install.json

BenHenning reviewed (commented):
Merge of upstream just involved a small changed to maven_install.json's input artifacts hash.

BenHenning committed:
Merge branch 'develop' into upgrade-to-newer-rules-kotlin

BenHenning committed:
Merge branch 'upgrade-to-newer-rules-kotlin' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Address review comment.

BenHenning committed:
Merge branch 'develop' into fix-build-time-kotlin-java-warnings

Conflicts:
	domain/src/main/java/org/oppia/android/domain/topic/PrimeTopicAssetsControllerImpl.kt

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

Conflicts:
	domain/BUILD.bazel
	third_party/versions.bzl

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning reviewed (commented):
Self-reviewed merges from upstream. Merges were clean, so nothing to double check.

adhiamboperes reviewed (approved):
This LGTM, thanks @BenHenning!

adhiamboperes commented:
Looks like `org.oppia.android.data.backends.gae.NetworkLoggingInterceptorTest > testLoggingInterceptor_makeNetworkCallWithInvalidUrl_failsAndCompletes` is flaky.

BenHenning committed:
Merge branch 'develop' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Merge branch 'develop' into fix-build-time-kotlin-java-warnings

BenHenning committed:
Merge branch 'fix-build-time-kotlin-java-warnings' into upgrade-to-kotlin1.6

BenHenning committed:
Add config for Kotlin warnings.

Add a config to allow warnings-as-errors to be disabled for local
development to make things easier. The default is still to treat them as
errors (and that will be the build configuration for CI, so PRs
shouldn't be mergeable with Kotlin build warnings).

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning reviewed (commented):
Reviewed latest merge changes--merge was clean.

BenHenning committed:
Address lint failures.

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning reviewed (commented):
Self-reviewed upstream merge (was clean).

BenHenning committed:
Merge branch 'develop' into upgrade-to-kotlin1.6

BenHenning committed:
Merge branch 'upgrade-to-kotlin1.6' into add-build-stats-ci-workflow

BenHenning reviewed (commented):
Self-reviewed latest changes (merge). They actually didn't seem 100% correct, but the actual WORKSPACE file is correct. I haven't seem that sort of thing come up before; it's a bit strange.

BenHenning commented:
Enabling auto-merge since there haven't been changes, the branch is up-to-date, and everything is resolved.

Note that the new workflow should kick in starting tomorrow after the PR is merged; I will keep an eye out to make sure it actually runs as expected.
